packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart

```diff
@@ -0,0 +1,150 @@
+import '../../base/error_handling_io.dart';
+import '../../base/file_system.dart';
+import '../../base/io.dart';
+import '../../base/logger.dart';
+import '../../flutter_manifest.dart';
+import '../build_system.dart';
+
+/// Applies a series of user-specified asset-transforming packages to an asset file.
+final class AssetTransformer {
+  AssetTransformer({
+    required ProcessManager processManager,
+    required Logger logger,
+    required FileSystem fileSystem,
+    required Artifacts artifacts,
+  })  : _processManager = processManager,
+        _logger = logger,
+        _fileSystem = fileSystem {
+          _dartBinary = artifacts.getArtifactPath(Artifact.engineDartBinary);
+        }
+
+  final ProcessManager _processManager;
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  late String _dartBinary;
+  final Random _random = Random();
+
+  /// The [Source] inputs that targets using this should depend on.
+  ///
+  /// See [Target.inputs].
+  static const List<Source> inputs = <Source>[
+    Source.pattern(
+      '{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/asset_transformer.dart',
+    ),
+  ];
+
+  /// Applies, in sequence, a list of transformers to an [asset] and then copies
+  /// the output to [outputPath].
+  ///
+  /// Returns `true` if successful and `false` otherwise.
+  Future<bool> transformAsset({
+    required File asset,
+    required String outputPath,
+    required String workingDirectory,
+    required List<AssetTransformerEntry> transformerEntries,
+  }) async {
+
+    String getTempFilePath() => '${_random.nextInt(1<<32)}${_fileSystem.path.extension(asset.path)}';
+
+    File tempInputFile = _fileSystem.systemTempDirectory.childFile(getTempFilePath());
+    await asset.copy(tempInputFile.path);
+    File tempOutputFile = _fileSystem.systemTempDirectory.childFile(getTempFilePath());
+
+    try {
+      for (final (int i, AssetTransformerEntry transformer) in transformerEntries.indexed) {
+        final _AssetTransformerFailure? transformerFailure = await _applyTransformer(
+          asset: tempInputFile,
+          output: tempOutputFile,
+          transformer: transformer,
+          workingDirectory: workingDirectory,
+        );
+
+        if (transformerFailure != null) {
+          _logger.printError(
+            'User-defined transformation of asset "${asset.path}" failed.\n'
+            '${transformerFailure.message}',
+          );
+          break;
+        }
+
+        ErrorHandlingFileSystem.deleteIfExists(tempInputFile);
+        if (i == transformerEntries.length - 1) {
+          await tempOutputFile.copy(outputPath);
+        } else {
+          tempInputFile = tempOutputFile;
+          tempOutputFile = _fileSystem.systemTempDirectory.childFile(getTempFilePath());
+        }
+      }
+    } finally {
+      ErrorHandlingFileSystem.deleteIfExists(tempInputFile);
+      ErrorHandlingFileSystem.deleteIfExists(tempOutputFile);
+    }
+
+    return true;
```

Comment:
what if we hit a `_AssetTransformerFailure`?

Reply:
Oh. This is handled at https://github.com/flutter/flutter/blob/16a21086d33ca96e21d830e64584d4d6a8d85655/packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart#L69-L75

but it should return `false` instead of breaking. This should be addressed now.

Reply:
:+1: